components/tidb_query_vec_executors/src/index_scan_executor.rs

```diff
@@ -232,59 +321,133 @@ impl IndexScanExecutorImpl {
+            }
+            let (value, remaining) = datum::split_datum(datum, false)?;
+            column.mut_raw().push(value);
+            *datum = remaining;
+        }
+        Ok(())
+    }
 
-            let handle_val = if tail_len >= 8 {
-                // This is a unique index, and we should look up PK handle in value.
-                self.decode_handle_from_value(&value[value.len() - 8..])?
-            } else {
-                // This is a normal index. The remaining payload part is the PK handle.
-                // Let's decode it and put in the column.
-                self.decode_handle_from_key(&key[key.len() - 9..])?
-            };
-            columns[self.columns_id_without_handle.len()]
-                .mut_decoded()
-                .push_int(Some(handle_val));
+    fn process_unique_common_handle_value(
+        &mut self,
+        key: &[u8],
+        value: &[u8],
+        columns: &mut LazyBatchColumnVec,
+    ) -> Result<()> {
+        let handle_len = ((value[2] as usize) << 8) + value[3] as usize;
+        let handle_end_offset = 4 + handle_len;
+
+        // If there are some restore data.
+        if handle_end_offset < value.len() {
+            let restore_values = &value[handle_end_offset..];
+            self.extract_columns_from_row_format(restore_values, columns)?;
+        } else {
+            // The datum payload part of the key.
+            let mut key_payload = &key[table::PREFIX_LEN + table::ID_LEN..];
+            Self::extract_columns_from_datum_format(
+                &mut key_payload,
+                &mut columns[0..self.columns_id_without_handle.len()],
+            )?;
+        }
+
+        let mut common_handle = &value[4..handle_end_offset];
+        Self::extract_columns_from_datum_format(
+            &mut common_handle,
+            &mut columns[self.columns_id_without_handle.len()..self.schema.len()],
+        )?;
+
+        Ok(())
+    }
+
+    fn process_normal_new_collation_value(
+        &mut self,
+        mut key: &[u8],
+        value: &[u8],
+        columns: &mut LazyBatchColumnVec,
+    ) -> Result<()> {
+        use DecodeHandleStrategy::*;
+        let tail_len = value[0] as usize;
+        let restore_values = &value[1..value.len() - tail_len];
+        self.extract_columns_from_row_format(restore_values, columns)?;
+
+        match self.decode_strategy {
+            NoDecode => {}
+            DecodeIntHandle if tail_len < 8 => {
+                key = &key[table::PREFIX_LEN + table::ID_LEN..];
+                datum::skip_n(&mut key, self.columns_id_without_handle.len())?;
+                let handle = self.decode_handle_from_key(key)?;
+                columns[self.columns_id_without_handle.len()]
+                    .mut_decoded()
+                    .push_int(Some(handle));
+            }
+            DecodeIntHandle => {
+                let handle = self.decode_handle_from_value(&value[value.len() - tail_len..])?;
```

Comment:
The previous behaviour is decode last 8 bytes, the current behaviour is decode first 8 bytes in the last tail_len bytes, they may be different. Please confirm about the bahaviour change.

Reply:
If there is no padding/untouch flag, then `tail_len` must be 8, otherwise, it will be 9.